src/dinit-monitor.cc

```diff
@@ -480,3 +558,95 @@ static bool load_service(int socknum, cpbuffer_t &rbuffer, const char *name, han
 
     return true;
 }
+
+static void request_environ(int socknum, cpbuffer_t &rbuffer)
+{
+    char c = (char)cp_cmd::LISTENENV;
+    write_all_x(socknum, &c, 1);
+
+    wait_for_reply(rbuffer, socknum);
+
+    cp_rply reply_pkt_h = (cp_rply)rbuffer[0];
+    if (reply_pkt_h != cp_rply::ACK) {
+        throw dinit_protocol_error();
+    }
+    rbuffer.consume(1);
+}
+
+static size_t get_allenv(int socknum, cpbuffer_t &rbuffer)
+{
+    char buf[2] = { (char)cp_cmd::GETALLENV, 0 };
+    write_all_x(socknum, buf, 2);
+
+    wait_for_reply(rbuffer, socknum);
+
+    cp_rply reply_pkt_h = (cp_rply)rbuffer[0];
+    if (reply_pkt_h != cp_rply::ALLENV) {
+        throw dinit_protocol_error();
+    }
+
+    // 1-byte packet header, then size_t
+    constexpr size_t allenv_hdr_size = 1 + sizeof(size_t);
+    rbuffer.fill_to(socknum, allenv_hdr_size);
+
+    size_t dsize;
+    rbuffer.extract(&dsize, 1, sizeof(dsize));
+    rbuffer.consume(allenv_hdr_size);
+
+    return dsize;
+}
+
+static bool issue_var(std::string &envar, const std::unordered_set<std::string> &varset, std::vector<stringview> &command_parts,
+        const char* str_set, const char* str_unset)
+{
+    auto eq = envar.find('=');
+    if (eq == envar.npos) {
+        /* unset */
+        eq = envar.size();
+    }
+    auto *sp = &envar[0];
+    sp[eq] = '\0';
+    if (varset.empty() || varset.find(sp) != varset.end()) {
+        if (eq == envar.size()) {
+            issue_command(sp, nullptr, str_unset, command_parts, true);
+        }
+        else {
+            issue_command(sp, &sp[eq + 1], str_set, command_parts, true);
+        }
+        return true;
+    }
+    return false;
+}
+
+static size_t read_and_issue(int socknum, cpbuffer_t &rbuffer, size_t dsz, const std::unordered_set<std::string> &varset,
```

Comment:
Could we name this `read_vars_and_issue` or something a little more specific? (I would tend to use `process_initial_vars` but whatever, just needs to be a bit more descriptive)

Reply:
done